stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+      // '.', 'e' or 'E'. In any case all that needs to be done here is
+      // to advance over the zero character and proceed to the optional
+      // fractional and exponential parts. If another digit follows then
+      // a parsing error will occur as expected, but implicitly because
+      // this function finishes with the parser positioned on a digit
+      // and not on a token expected after a number like ',', ']', '}' or
+      // EOF.
+      addCharFromCodePoint(c, buffer)
+      c = next(parserState)
+    },
+    x when x >= 0x31 && x <= 0x39 => { // '1'..'9'
+      for (;;) {
+        addCharFromCodePoint(c, buffer)
+        c = next(parserState)
+        if (c < 0x30 || c > 0x39) {
+          break
+        }
+      }
+      void
+    },
+    unexpectedCodePoint => {
+      // The integer part of the number has to have at least one digit.
+      // JSON doesn't allow numbers starting with decimal separator like ".1".
+      let detail = "expected a decimal digit, found " ++
+        formatCodePointOrEOF(unexpectedCodePoint)
+      return Err(buildUnexpectedTokenError(parserState, detail))
+    },
+  }
+  // Optional fractional part of the number.
+  if (c == 0x2E) { // '.'
+    isFloat = true
+    Buffer.addChar('.', buffer)
+    c = next(parserState)
+    let mut hasHitDigit = false
+    for (; c >= 0x30 && c <= 0x39;) {
+      hasHitDigit = true
+      addCharFromCodePoint(c, buffer)
+      c = next(parserState)
+    }
+    if (!hasHitDigit)
+      return Err(
+        buildUnexpectedTokenError(
+          parserState,
+          "exponent part is missing in number"
+        )
+      )
+  }
+  // Optional exponential part of the number.
+  if (c == 0x65 || c == 0x45) { // 'e' or 'E'
+    isFloat = true
+    Buffer.addChar('e', buffer)
+    c = next(parserState)
+    // can start with optional plus or minus sign
+    match (c) {
+      0x2D => { // '-'
+        c = next(parserState)
+        Buffer.addChar('-', buffer)
+      },
+      0x2B => { // '+'
+        c = next(parserState)
+      },
+      _ => void,
+    }
+    // followed by one or more digits (0-9)
+    let mut hasHitDigit = false
+    for (; c >= 0x30 && c <= 0x39;) {
+      hasHitDigit = true
+      addCharFromCodePoint(c, buffer)
+      c = next(parserState)
+    }
+    if (!hasHitDigit)
+      return Err(
+        buildUnexpectedTokenError(
+          parserState,
+          "exponent part is missing in number"
+        )
+      )
+
+    void
+  }
+  // Note that unlike all other JSON value types there's no explicit ending
+  // character like ('"' for strings, ']' for arrays,'}' for objects etc). We
+  // just leave the parser state at current position and the reading of next
+  // token will succeed or fail, but number parsing just ends here.
+  let result = match (isFloat) {
+    false => atoiFast(buffer),
+    true => {
+      let str = Buffer.toString(buffer)
+      match (Number.parseFloat(str)) {
+        Err(err) => fail "Impossible",
```

Comment:
This needs to be a better fail message so we can link back if it actually happens. `Impossible: JSON parse float on a non-float` or something similar